src/api/schema/log_events/mod.rs

```diff
@@ -0,0 +1,170 @@
+    let (mut log_tx, log_rx) = mpsc::channel::<Vec<LogEventPayload>>(10);
+
+    tokio::spawn(async move {
+        // Create a tap sink. When this drops out of scope, clean up will be performed on the
+        // event handlers and topology observation that the tap sink provides.
+        let _tap_sink = TapSink::new(watch_rx, tap_tx, &component_names);
+
+        // A tick interval to represent when to 'cut' the results back to the client.
+        let mut interval = time::interval(time::Duration::from_millis(interval));
+
+        // Collect a vector of results, with a capacity of `limit`. As new `LogEvent`s come in,
+        // they will be sampled and added to results.
+        let mut results = Vec::<LogEventPayload>::with_capacity(limit);
+
+        // Random number generator to allow for sampling. Speed trumps cryptographic security here.
+        // The RNG must be Send + Sync to use with the `select!` loop below.
+        let mut rng = SmallRng::from_entropy();
+
+        // Keep a count of the batch size, which will be used as a seed for random eviction
+        // per the sampling strategy used below.
+        let mut batch = 0;
+
+        loop {
+            select! {
+                // Process `TapPayload`s. A tap payload could contain a `LogEvent` or a notification.
+                // Notifications are emitted immediately; log events buffer until the next `interval`.
+                Some(tap) = tap_rx.next() => {
+                    let tap = tap.into();
+
+                    // Emit notifications immediately; these don't count as a 'batch'.
+                    if let LogEventPayload::Notification(_) = tap {
+                        // If an error occurs when sending, the subscription has likely gone
+                        // away. Break the loop to terminate the thread.
+                        if log_tx.send(vec![tap]).await.is_err() {
+                            break;
+                        }
+                    } else {
+                        // A simple implementation of "Algorithm R" per
+                        // https://en.wikipedia.org/wiki/Reservoir_sampling. As we're unable to
+                        // pluck the nth result, this is chosen over the more optimal "Algorithm L"
+                        // since discarding results isn't an option.
+                        if limit > results.len() {
+                            results.push(tap);
+                        } else {
+                            let random_number = rng.gen_range(0..batch);
+                            if random_number < results.len() {
+                                results[random_number] = tap;
+                            }
+                        }
+                        // Increment the batch count, to be used for the next Algo R loop
+                        batch += 1;
+                    }
+                }
+                _ = interval.tick() => {
```

Comment:
This is mostly for my edification, but could this arm of the `select!` be starved if `tap_rx.next()` always produces an event. It seems like yes looking at: https://docs.rs/tokio/0.3.7/tokio/macro.select.html

Maybe we could move this arm up so that it always executes on the interval?

Reply:
I don't _think_ that should work in practice, due to each loop selecting a branch at random:

> select! randomly picks a branch to check first. This provides some level of fairness when calling select! in a loop with branches that are always ready.

If `.tick` is ready from the previous loop, there's a 50% chance it'll be selected in the subsequent, so it should be at or close to the elapsed interval +/- the cost of processing a single event, plus the probability of subsequent non-selection.

There's a merged issue at https://github.com/tokio-rs/tokio/pull/3603 for introducing a `biased!` option to prefer one over the other, but random selection against two branches should be far off in practice. That seems like a pattern worth adopting when it's available, in any case.

Reply:
Aha, you are right, I missed the note about fairness on that page. I was misinterpreting "Wait on multiple concurrent branches, returning when the first branch completes, cancelling the remaining branches." line appearing above it.

👍 to using `biased!` when it is available to prefer the interval branch.